### PR TITLE
Allow using const values in agent macro

### PR DESCRIPTION
**Pull request by Tortar**
<del>This should make it possible to use default values addressing https://github.com/JuliaDynamics/Agents.jl/issues/709 , let me know what do you think.  </del> Experimenting a bit, there shouldn't be any perf regression applying @kwdef as default

<del>(I also put in the api page of the documentation the replicate! function, forgot to do that in the other pr :D) </del>

edit: now it also includes the possibility to declare constants, which should help performance, but need some good tests

### CODE REVIEW

**Comment by Tortar** [comment]
Implemented the first way described in #728 

```julia
using Agents

@agent A NoSpaceAgent begin
    f1::Int
    f2::Int
    consts = (:f1,)
end

agent = A(1, 3, 5)
agent.f1 = 10
```

returns

```julia
ERROR: setfield!: const field .f1 of type A cannot be changed
Stacktrace:
 [1] setproperty!(x::A, f::Symbol, v::Int64)
   @ Base .\Base.jl:38
 [2] top-level scope
   @ REPL[4]:1
```

**Comment by Tortar** [comment]
this should be it, all new tests passes

**Comment by Datseris** [comment]
does it allow inhereting from a type defined with constants? if we change `NoSpaceAgent` to have `const id::Int`, does `@agent` work with inhereting from `NoSpaceAgent`?

**Comment by Tortar** [comment]
Yes it allows it, any const will be propagated to any inheriting new type, this is so because we actually check if there are any const field in the base type and then add it as a const also in the new type

**Comment by Datseris** [comment]
ok great!

File: src/core/agents.jl [hunk 1]
```diff
@@ -28,9 +28,7 @@ Define an agent struct which includes all fields that `AnotherAgentType` has,
 as well as any additional ones the user may provide via the `begin` block.
 See below for examples.
 
-Using `@agent` is the recommended way to create agent types for Agents.jl,
-however keep in mind that the macro (currently) doesn't work with `const` 
-declarations in individual fields (for Julia v1.8+).
+Using `@agent` is the recommended way to create agent types for Agents.jl.
```
Comment (Datseris) [inline src/core/agents.jl:31]:
```suggestion
Using `@agent` is the only supported way to create agent types for Agents.jl.
```

**Comment by Datseris** [comment]
why isn't it possible to do things the native julia way and allow prefacing with `const` a constant field?

**Comment by Tortar** [comment]
yes, a field name consts is disallowed, there is no way to have const inside the agent macro unfortunately due to parsing errors, see https://github.com/JuliaLang/julia/issues/45024

**Comment by Tortar** [comment]
it's a bit hacky yes but it's a way not to redefine the agent macro as was done in https://github.com/JuliaDynamics/Agents.jl/issues/728, we could follow that path alternatevely but this means that every code will break :D

**Comment by Datseris** [comment]
but the julia issue you cite is fixed...?

**Comment by Datseris** [comment]
Okay, I see, there is no way to have `const x = 1`. Okay. I think your solution with the extra field is the only option then.

But, rename it to `constants`. and document it appropriately.

@Tortar this is a big change. please do not tag anything after mergfing the const and keyword improvements. i want to read the final docstring and changelog and give the ok for a new tag.

**Comment by Tortar** [comment]
yes, but the fix is disallowing the construct :D

Il Sab 22 Lug 2023, 11:59 George Datseris ***@***.***> ha
scritto:

**Comment by Tortar** [comment]
sure, I actually have another important change which I would like to finish up, we should probably tag them all together

**Comment by Tortar** [comment]
@Datseris I completed the higlighting of the changes reverting some of the changed lines in https://github.com/JuliaDynamics/Agents.jl/pull/732/files , I also declared the `id` field of `NoSpaceAgent` as constant

File: src/core/agents.jl [hunk 2]
```diff
@@ -103,6 +101,19 @@ Person(id = 1, pos = (1, 1), moneyz = 2000)
 # new age value
 Person(1, (1, 1), 40, 2000)
 ```
+It is also possible to specify that some fields are immutable
```
Comment (Datseris) [inline src/core/agents.jl:104]:
```suggestion
It is also possible to specify that some fields are immutable (constants)
```

File: test/runtests.jl [hunk 3]
```diff
@@ -77,6 +77,26 @@ Agent8(id, pos; f1, f2) = Agent8(id, pos, f1, f2)
     f3::Float64 = 3.0
 end
 
+@agent Agent10 NoSpaceAgent begin
```
Comment (Datseris) [inline test/runtests.jl:80]: please do not define agent types outside the test they are needed. I know this is what we do in the tests at the moment very often but it is very bad practice that we are slowly fixing.

Comment (Datseris) [reply test/runtests.jl:80]: see https://github.com/JuliaDynamics/Agents.jl/issues/634

**Review by Datseris — APPROVED** [review]
Sounds great. Simple comments to address and then please merge (BUT NNOT TAG!!!)

**Comment by Tortar** [comment]
Merging! (Not tagging :D)

**Pull request merged**